### PR TITLE
refactor(batch): use futures-async-stream to implement hash agg

### LINKED ISSUE #1936: Use futures-async-stream to refactor  HashAggExecutor.

### PR DESCRIPTION
**Pull request by StrikeW**
## What's changed and what's your intention?

- use futures-async-stream to implement batch hash agg
- update unit tests

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close #1936

### CODE REVIEW

File: src/batch/src/executor2/hash_agg.rs [hunk 1]
```diff
@@ -192,66 +207,50 @@ impl<K: HashKey + Send + Sync> Executor for HashAggExecutor<K> {
...
+
+                let mut agg_builders = self
+                    .agg_factories
+                    .iter()
+                    .map(|agg_factory| {
+                        agg_factory
+                            .get_return_type()
+                            .create_array_builder(cardinality)
+                    })
+                    .collect::<Result<Vec<_>>>()?;
```
Comment (fuyufjh) [inline src/batch/src/executor2/hash_agg.rs:228]: Is it possible to reuse these array builders (`group_builders` and `agg_builders`) across the loops?

**Review by fuyufjh — APPROVED** [review]
LGTM for rest

File: src/batch/src/executor2/hash_agg.rs (hunk 1 above)
Comment (StrikeW) [reply src/batch/src/executor2/hash_agg.rs:228]: I also notice that. But if we move the builders outside the loop, it will incur some ownership problems.

Comment (StrikeW) [reply src/batch/src/executor2/hash_agg.rs:228]: And each time we build an output chunk, we need to deserialize the hash key into builders, seems that we cannot reuse them.

File: src/batch/src/executor2/hash_agg.rs [hunk 2]
```diff
@@ -128,13 +129,13 @@ impl BoxedExecutorBuilder for HashAggExecutorBuilder {
...
+pub(crate) struct HashAggExecutor2<K> {
     /// factories to construct aggregator for each groups
     agg_factories: Vec<AggStateFactory>,
     /// Column indexes of keys that specify a group
     group_key_columns: Vec<usize>,
     /// child executor
-    child: BoxedExecutor,
+    child: BoxedExecutor2,
     /// hash map for each agg groups
     groups: AggHashMap<K>,
```
Comment (liurenjie1024) [inline src/batch/src/executor2/hash_agg.rs:140]: We nolonger need these to store as fields.

File: src/batch/src/executor2/hash_agg.rs [hunk 3]
```diff
@@ -192,66 +207,50 @@ impl<K: HashKey + Send + Sync> Executor for HashAggExecutor<K> {
...
+                    .agg_factories
+                    .iter()
+                    .map(|agg_factory| {
+                        agg_factory
+                            .get_return_type()
+                            .create_array_builder(cardinality)
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                let mut has_next = false;
```
Comment (liurenjie1024) [inline src/batch/src/executor2/hash_agg.rs:230]: We don't need this?

File: src/batch/src/executor2/hash_agg.rs (hunk 2 above)
Comment (StrikeW) [reply src/batch/src/executor2/hash_agg.rs]: fixed

File: src/batch/src/executor2/hash_agg.rs [hunk 4]
```diff
@@ -205,7 +205,7 @@ impl<K: HashKey + Send + Sync> HashAggExecutor2<K> {
         }
 
         // the aggregated result set
-        let mut result = Some(mem::take(&mut groups).into_iter());
+        let mut result = Some(groups.into_iter());
```
Comment (liurenjie1024) [inline src/batch/src/executor2/hash_agg.rs:208]: Why not just `let mut result = groups.into_iter()`

**Pull request merged**